### ISSUE TITLE
VPN seed error breaks communication to API server

### ISSUE THREAD

**Issue by tobiscr** [gardener/gardener#6588]
**How to categorize this issue?**

/area networking
/kind bug

**What happened**:

We are facing in our pipelines frequently happening issues which seem to be related to the vpn-seed server communication:

```
Error from server: error dialing backend: proxy error from vpn-seed-server:9443 while dialing 10.250.0.2:10250, code 503: 503 Service Unavailable
```

When the error message is reported, the communication to the API server become unreliable and stops working either immediately or with a delay between 30 sec - 1 min.

Example for an immediately failed pipeline:
* https://storage.googleapis.com/kyma-prow-logs/pr-logs/pull/kyma-project_control-plane/1986/pre-main-control-plane-reconciler-e2e/1563149721512448000/build-log.txt
* https://storage.googleapis.com/kyma-prow-logs/pr-logs/pull/kyma-project_control-plane/1666/pre-main-control-plane-reconciler-upgrade-kyma2-latest-to-main-gardener/1562728027379470336/build-log.txt
* https://storage.googleapis.com/kyma-prow-logs/pr-logs/pull/kyma-project_control-plane/1986/pre-main-control-plane-reconciler-e2e/1562787208908247040/build-log.txt

Example for a pipeline which went into communcation issues a few seconds after the error occured:
* https://status.build.kyma-project.io/view/gs/kyma-prow-logs/pr-logs/pull/kyma-project_control-plane/1986/pre-main-control-plane-reconciler-upgrade-kyma2-latest-to-main-gardener/1563084664967532544
* https://status.build.kyma-project.io/view/gs/kyma-prow-logs/pr-logs/pull/kyma-project_control-plane/1960/pre-main-control-plane-reconciler-upgrade-kyma2-latest-to-main-gardener/1562822462784147456
* https://status.build.kyma-project.io/view/gs/kyma-prow-logs/pr-logs/pull/kyma-project_control-plane/1666/pre-main-control-plane-reconciler-upgrade-kyma2-latest-to-main-gardener/1562808546834255872

This problem happens frequently (but not always) and is currently one of the main reasons for red pipelines on our side.

**What you expected to happen**:

Communciation to API server is stable and not behaving flacky.

**How to reproduce it (as minimally and precisely as possible)**:

Randomly happening, see history of the example pipelines. You usually find in at least 25% of the failed pipelines report this misbehaviour. 

**Anything else we need to know?**:

If you need further information, please don't hesitate to get in touch with me.

**Environment**:

- Gardener version: 
- Kubernetes version (use `kubectl version`):  1.23.6

       "major": "1",
         "minor": "23",
         "gitVersion": "v1.23.6",
         "gitCommit": "ad3338546da947756e8a88aa6822e9c11e7eac22",
         "gitTreeState": "clean",
         "buildDate": "2022-04-14T08:43:11Z",
         "goVersion": "go1.17.9",
         "compiler": "gc",
         "platform": "linux/amd64"

- Cloud provider or hardware configuration: GCP
- Others:

**Comment by rfranzke** [comment]
cc @ScheererJ @DockToFuture

**Comment by dee0sap** [comment]
My team has experienced this as well recently however it was outside of Kyma.

It happened 

- A few times In this pipeline https://sacservices.jaas-gcp.cloud.sap.corp/blue/organizations/jenkins/dmi-multicloud-service-provisioner/detail/PR-122/13/pipeline/2111/.    We added some retry logic to get around it.
- When invoking kubectl logs 

A teammate noticed the following sequence of messages in the log of a vpn-shoot pod in the kube-system namespace.  The timestamps around the time the connection dropped.    No idea if they point to the root cause or if they are just a secondary symptom.

2022-08-24 03:00:33 Connection reset, restarting [0]
2022-08-24 03:00:33 /sbin/ip addr del dev tun0 192.168.123.2/24
2022-08-24 03:00:33 SIGTERM[soft,connection-reset] received, process exiting
2022-08-24 03:00:34 OpenVPN 2.5.6 x86_64-alpine-linux-musl [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [MH/PKTINFO] [AEAD] built on Mar 24 2022

**Comment by ScheererJ** [comment]
I may state the obvious, but the current vpn solution used by Gardener is not highly available as of today. There is exactly one connection between the endpoint in the control plane and the cluster. Both sides can break, e.g. due to autoscaling or a node going down. When this happens the connection will be re-established, but this may take long enough that some requests will fail.
High availability for the vpn solution is on the roadmap, but without a date being set.
In case you want me to investigate on a concrete cluster please share the details in separate communication.

**Comment by dee0sap** [comment]
@ScheererJ 
Thanks for that info.
At least in the case of my team, we have only seen this recently and only with transient clusters.   e.g.  Clusters created by a CD pipeline for integration testing.
The latter detail makes it difficult to provide access as generally the cluster is gone.

The pipeline does dump the manifests of all k8s objects and it gathers nearly all logs from the cluster before it is deleted.    Would this info be useful to you?   Or perhaps there are some specific, additional, things the pipeline could collect that would be useful?

**Comment by ScheererJ** [comment]
@dee0sap
Unfortunately, the kubernetes manifests do not help in this regard. As mentioned before, I have seen several cases where either autoscaling or simply node rotation caused vpn outages as the vpn pods needed to be rescheduled/restarted. You can easily rule this out if you look at the age of the `vpn-shoot` pod in the shoot cluster and the age of the `vpn-seed-server` pod in the control plane. However, depending on your setup you may not have full access to your control plane.

**Comment by dee0sap** [comment]
Thanks @ScheererJ 
Our dump of information includes manifests from pods, events, and logs
So we'll have the age of the vpn-shoot pod in the shoot cluster.

Am not sure if we collect anything that will provide the age of the vpn-seed-server.   I'll look into that.

Please let me know if there is any other data my team should try to collect.

**Comment by ScheererJ** [comment]
@dee0sap
In case there are node changes it might also be relevant, i.e. nodes going away and being replaced.
Some of the data concerning the control plane, e.g. related to `vpn-seed-server` pod may be available indirectly via observability metrics, but it might be a pain to get this info out of prometheus.

**Comment by tobiscr** [comment]
Sorry for the late update. 

We are still facing VPN issues, but a retry solves the problem in most cases. We had just 1 cluster in last few days which was consistantly blocking any API server communication (https://storage.googleapis.com/kyma-prow-logs/pr-logs/pull/kyma-project_control-plane/2031/pre-main-control-plane-reconciler-upgrade-kyma2-latest-to-main-gardener/1568212161677758464/build-log.txt).

We are still trying to provide such a cluster which is permanently failing.

**Comment by tobiscr** [comment]
The issue happened today again and is not solved:

https://status.build.kyma-project.io/view/gs/kyma-prow-logs/pr-logs/pull/kyma-project_control-plane/2076/pre-main-control-plane-reconciler-upgrade-kyma2-latest-to-main-gardener/1574721267871780864

 We are still trying to provide a cluster for further analysis. Please don't close this ticket until we got one defective cluster in place.

**Comment by ScheererJ** [comment]
/remove-lifecycle stale

Please note that we have meanwhile implemented a highly available VPN with #6978. You can activate its usage as described in the pull request if your seed cluster supports it. It resolves several error cases of the old solution.

/close
